src/compiler.ts

```diff
@@ -4617,79 +4619,205 @@ export class Compiler extends DiagnosticEmitter {
           return this.module.unreachable();
         }
 
-        let targetType = leftType;
-        let instance: Function | null;
-
-        // Mathf.pow if lhs is f32 (result is f32)
-        if (this.currentType.kind == TypeKind.F32) {
-          rightExpr = this.compileExpression(right, Type.f32, Constraints.CONV_IMPLICIT);
+        if (compound) {
+          leftExpr = this.ensureSmallIntegerWrap(leftExpr, leftType);
+          rightExpr = this.compileExpression(right, leftType, Constraints.CONV_IMPLICIT);
           rightType = this.currentType;
-          instance = this.f32PowInstance;
-          if (!instance) {
-            let namespace = this.program.lookupGlobal(CommonNames.Mathf);
-            if (!namespace) {
-              this.error(
-                DiagnosticCode.Cannot_find_name_0,
-                expression.range, "Mathf"
-              );
+        } else {
+          rightExpr = this.compileExpression(right, leftType);
+          rightType = this.currentType;
+          commonType = Type.commonDenominator(leftType, rightType, false);
+          if (commonType) {
+            leftExpr = this.convertExpression(leftExpr,
+              leftType, commonType,
+              false, true, // !
+              left
+            );
+            leftType = commonType;
+            rightExpr = this.convertExpression(rightExpr,
+              rightType, commonType,
+              false, true, // !
+              right
+            );
+            rightType = commonType;
+          } else {
+            this.error(
+              DiagnosticCode.Operator_0_cannot_be_applied_to_types_1_and_2,
+              expression.range, "**", leftType.toString(), rightType.toString()
+            );
+            this.currentType = contextualType;
+            return module.unreachable();
+          }
+        }
+
+        let instance: Function | null;
+        switch (this.currentType.kind) {
```

Comment:
Might be clearer to switch over `commonType` here, which is the same. Just needs to be set in the `if (compound) {` case as well then.

Reply:
done!